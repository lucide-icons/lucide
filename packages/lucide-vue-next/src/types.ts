import type { FunctionalComponent, SVGAttributes } from 'vue';

export interface LucideProps extends Partial<SVGAttributes> {
  size?: 24 | number;
  strokeWidth?: number | string;
  absoluteStrokeWidth?: boolean;
}

export type IconNode = [elementName: string, attrs: Record<string, string>][];
export type LucideIcon = FunctionalComponent<LucideProps>;

// Legacy exports
export type SVGProps = LucideProps;
