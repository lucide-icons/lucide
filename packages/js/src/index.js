import icons from './icons';
import toSvg from './to-svg';
import replace from './replace';

module.exports = { icons, toSvg, replace };
