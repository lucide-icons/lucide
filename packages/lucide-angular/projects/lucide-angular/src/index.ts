
export * from './lib/lucide-angular.component'
export * from './lib/lucide-angular.module';
export * from './helpers/create-element'

export * from './icons'
export * as icons from './icons';
