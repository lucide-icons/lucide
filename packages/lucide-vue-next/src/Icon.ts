import { type FunctionalComponent, h } from 'vue';
import { mergeClasses, toKebabCase } from '@lucide/shared';
import defaultAttributes from './defaultAttributes';
import { IconNode, LucideProps } from './types';

interface IconProps {
  iconNode: IconNode;
  name: string;
}

const Icon: FunctionalComponent<LucideProps & IconProps> = (
  { size, strokeWidth = 2, absoluteStrokeWidth, color, iconNode, name, class: classes, ...props },
  { slots },
) => {
  return h(
    'svg',
    {
      ...defaultAttributes,
      width: size || defaultAttributes.width,
      height: size || defaultAttributes.height,
      stroke: color || defaultAttributes.stroke,
      'stroke-width': absoluteStrokeWidth ? (Number(strokeWidth) * 24) / Number(size) : strokeWidth,
      class: ['lucide', `lucide-${toKebabCase(name ?? 'icon')}`],
      ...props,
    },
    [...iconNode.map((child) => h(...child)), ...(slots.default ? [slots.default()] : [])],
  );
};

export default Icon;
