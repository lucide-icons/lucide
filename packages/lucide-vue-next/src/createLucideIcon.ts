import { h } from 'vue';
import type { SVGAttributes, FunctionalComponent } from 'vue';
import defaultAttributes from './defaultAttributes';

// Create interface extending SVGAttributes
export interface SVGProps extends Partial<SVGAttributes> {
  size?: 24 | number
}


type IconNode = [elementName: string, attrs: Record<string, string>][]

/**
 * Converts string to KebabCase
 * Copied from scripts/helper. If anyone knows how to properly import it here
 * then please fix it.
 *
 * @param {string} string
 * @returns {string} A kebabized string
 */
export const toKebabCase = (string: string) => string.replace(/([a-z0-9])([A-Z])/g, '$1-$2').toLowerCase();

const createLucideIcon = (iconName: string, iconNode: IconNode): FunctionalComponent<SVGProps> => (
  { size, color, ...props }, // props
  { attrs, slots } // context
  ) => {
  return h(
    'svg',
    {
      ...defaultAttributes,
      width: size || defaultAttributes.width,
      height: size || defaultAttributes.height,
      stroke: color || defaultAttributes.stroke,
      ...attrs,
      class: ['lucide', `lucide-${toKebabCase(iconName)}`, attrs?.class || ''],
      ...props,
    },
    [
      ...iconNode.map(child => h(...child)),
      ...(slots.default ? [slots.default()] : [])
    ],
  );
};

export default createLucideIcon;
