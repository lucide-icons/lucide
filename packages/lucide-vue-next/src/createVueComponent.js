import { h } from 'vue';
import defaultAttributes from './defaultAttributes';

const createVueComponent = (iconName, iconNode) => (props, context) =>
  h(
    'svg',
    {
      ...defaultAttributes,
      ...{
        width: props.size || defaultAttributes.width,
        height: props.size || defaultAttributes.height,
      },
      ...context.attrs,
      ...props,
    },
    iconNode.map(child => h(...child)),
  );

export default createVueComponent;
