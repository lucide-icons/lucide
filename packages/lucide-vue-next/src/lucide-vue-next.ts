export * from './icons';
export * as icons from './icons';
export * from './aliases';

export { default as createLucideIcon } from './createLucideIcon';
export { default as useIconComponent } from './useIconComponent';
