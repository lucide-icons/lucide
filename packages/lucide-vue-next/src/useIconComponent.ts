import createLucideIcon from './createLucideIcon';
import { LucideIcon, IconNode } from './types';

type CamelToPascal<T extends string> = T extends `${infer FirstChar}${infer Rest}`
  ? `${Capitalize<FirstChar>}${Rest}`
  : never;

type ComponentList<T> = {
  [Prop in keyof T as CamelToPascal<Prop & string>]: LucideIcon;
};

export const toPascalCase = <T extends string>(string: T): CamelToPascal<T> => {
  const camelCase = string.replace(/^([A-Z])|[\s-_]+(\w)/g, (match, p1, p2) =>
    p2 ? p2.toUpperCase() : p1.toLowerCase(),
  );

  return (camelCase.charAt(0).toUpperCase() + camelCase.slice(1)) as CamelToPascal<T>;
};

const useIconComponent = <Icons extends Record<string, IconNode>>(icons: Icons) => {
  if (typeof icons !== 'object') {
    throw new Error('[lucide-vue-next]: useIconComponent expects an object as argument');
  }

  const iconNodeEntries = Object.entries(icons);

  const iconNodesHasCorrectType = iconNodeEntries.every(([, iconNode]) => Array.isArray(iconNode));

  if (!iconNodesHasCorrectType) {
    throw new Error('[lucide-vue-next]: Passed icons object has incorrect type');
  }

  const iconComponents = iconNodeEntries.reduce((acc, [iconName, iconNode]) => {
    const componentName = toPascalCase(iconName) as keyof ComponentList<Icons>;

    acc[componentName] = createLucideIcon(
      componentName as string,
      iconNode,
    ) as ComponentList<Icons>[typeof componentName];

    return acc;
  }, {} as ComponentList<Icons>);

  return iconComponents;
};

export default useIconComponent;
