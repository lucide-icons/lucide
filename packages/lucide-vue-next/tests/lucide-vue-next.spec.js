import {render, fireEvent} from '@testing-library/vue'
import { Smile } from '../src/icons'

describe('Using lucide icon components', () => {
  it('should render an component', () => {
    const { container } = render(Smile)
    expect(container).toMatchSnapshot();
  });

  it('should adjust the size, stroke color and stroke width', () => {
    const {container} = render(Smile, {
      props: {
        size: 48,
        color: 'red',
        'stroke-width': 4
      }
    })

    const [icon] = document.getElementsByClassName('lucide');

    expect(icon.getAttribute('width')).toBe('48')
    expect(icon.getAttribute('stroke')).toBe('red')
    expect(icon.getAttribute('stroke-width')).toBe('4')

    expect(container).toMatchSnapshot();
  });


  it('should add a class to the element', () => {
    const {container} = render(Smile, {
      attrs: {
        class: "my-icon"
      }
    })

    expect(container).toMatchSnapshot();

    const [icon] = document.getElementsByClassName('my-icon');

    console.log(icon);

    expect(icon).toHaveClass('my-icon')
  });

  it('should add a style attribute to the element', () => {
    const {container} = render(Smile, {
      attrs: {
        style: 'position: absolute',
      }
    })

    expect(container).toMatchSnapshot();

    const [icon] = document.getElementsByClassName('lucide');

    expect(icon).toHaveStyle({ position: 'absolute' })
  });

  it('should call the onClick event', async () => {
    const onClick = jest.fn()
    render(Smile, {
      attrs: {
        onClick,
      }
    })

    const [icon] = document.getElementsByClassName('lucide');

    await fireEvent.click(icon)

    expect(onClick).toHaveBeenCalled()
  });

  it('should pass children to the icon slot', () => {
    const testText = 'Hello World'
    const template = {
      name: 'Stub',
      template: `<text>${testText}</text>`
    }
    const { getByText, container } = render(Smile, {
      slots: {
        default: template
      }
    })

    const textElement = getByText(testText)

    expect(textElement).toBeInTheDocument()
    expect(container).toMatchSnapshot();
  });
});
