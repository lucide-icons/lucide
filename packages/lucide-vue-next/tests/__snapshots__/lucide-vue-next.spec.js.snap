// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Using lucide icon components should add a class to the element 1`] = `
VueWrapper {
  "__app": Object {
    "_component": Object {
      "__emits": Object {},
      "__props": Array [
        Object {},
        Array [],
      ],
      "name": "VTU_ROOT",
      "render": [Function],
    },
    "_container": <div
      data-v-app=""
    >
      <svg
        class="lucide-icon my-icon"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="12"
          cy="12"
          r="10"
        />
        <path
          d="M8 14s1.5 2 4 2 4-2 4-2"
        />
        <line
          x1="9"
          x2="9.01"
          y1="9"
          y2="9"
        />
        <line
          x1="15"
          x2="15.01"
          y1="9"
          y2="9"
        />
      </svg>
    </div>,
    "_context": Object {
      "app": [Circular],
      "components": Object {
        "transition": Object {
          "name": "transition",
          "props": undefined,
          "render": [Function],
        },
        "transition-group": Object {
          "name": "transition-group",
          "props": undefined,
          "render": [Function],
        },
      },
      "config": Object {
        "errorHandler": undefined,
        "globalProperties": Object {},
        "isCustomElement": [Function],
        "isNativeTag": [Function],
        "optionMergeStrategies": Object {},
        "performance": false,
        "warnHandler": undefined,
      },
      "directives": Object {},
      "mixins": Array [
        Object {
          "__emits": null,
          "__props": Array [],
          "beforeCreate": [Function],
        },
      ],
      "provides": Object {},
      "reload": [Function],
    },
    "_props": null,
    "_uid": 2,
    "component": [Function],
    "config": Object {
      "errorHandler": undefined,
      "globalProperties": Object {},
      "isCustomElement": [Function],
      "isNativeTag": [Function],
      "optionMergeStrategies": Object {},
      "performance": false,
      "warnHandler": undefined,
    },
    "directive": [Function],
    "mixin": [Function],
    "mount": [Function],
    "provide": [Function],
    "unmount": [Function],
    "use": [Function],
    "version": "3.0.6",
  },
  "__setProps": [Function],
  "componentVM": Object {
    "hasOwnProperty": [Function],
  },
  "rootVM": Object {},
  "wrapperElement": <svg
    class="lucide-icon my-icon"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      cx="12"
      cy="12"
      r="10"
    />
    <path
      d="M8 14s1.5 2 4 2 4-2 4-2"
    />
    <line
      x1="9"
      x2="9.01"
      y1="9"
      y2="9"
    />
    <line
      x1="15"
      x2="15.01"
      y1="9"
      y2="9"
    />
  </svg>,
}
`;

exports[`Using lucide icon components should add a style attribute to the element 1`] = `
VueWrapper {
  "__app": Object {
    "_component": Object {
      "__emits": Object {},
      "__props": Array [
        Object {},
        Array [],
      ],
      "name": "VTU_ROOT",
      "render": [Function],
    },
    "_container": <div
      data-v-app=""
    >
      <svg
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        style="position: absolute;"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="12"
          cy="12"
          r="10"
        />
        <path
          d="M8 14s1.5 2 4 2 4-2 4-2"
        />
        <line
          x1="9"
          x2="9.01"
          y1="9"
          y2="9"
        />
        <line
          x1="15"
          x2="15.01"
          y1="9"
          y2="9"
        />
      </svg>
    </div>,
    "_context": Object {
      "app": [Circular],
      "components": Object {
        "transition": Object {
          "name": "transition",
          "props": undefined,
          "render": [Function],
        },
        "transition-group": Object {
          "name": "transition-group",
          "props": undefined,
          "render": [Function],
        },
      },
      "config": Object {
        "errorHandler": undefined,
        "globalProperties": Object {},
        "isCustomElement": [Function],
        "isNativeTag": [Function],
        "optionMergeStrategies": Object {},
        "performance": false,
        "warnHandler": undefined,
      },
      "directives": Object {},
      "mixins": Array [
        Object {
          "__emits": null,
          "__props": Array [],
          "beforeCreate": [Function],
        },
      ],
      "provides": Object {},
      "reload": [Function],
    },
    "_props": null,
    "_uid": 3,
    "component": [Function],
    "config": Object {
      "errorHandler": undefined,
      "globalProperties": Object {},
      "isCustomElement": [Function],
      "isNativeTag": [Function],
      "optionMergeStrategies": Object {},
      "performance": false,
      "warnHandler": undefined,
    },
    "directive": [Function],
    "mixin": [Function],
    "mount": [Function],
    "provide": [Function],
    "unmount": [Function],
    "use": [Function],
    "version": "3.0.6",
  },
  "__setProps": [Function],
  "componentVM": Object {
    "hasOwnProperty": [Function],
  },
  "rootVM": Object {},
  "wrapperElement": <svg
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    style="position: absolute;"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      cx="12"
      cy="12"
      r="10"
    />
    <path
      d="M8 14s1.5 2 4 2 4-2 4-2"
    />
    <line
      x1="9"
      x2="9.01"
      y1="9"
      y2="9"
    />
    <line
      x1="15"
      x2="15.01"
      y1="9"
      y2="9"
    />
  </svg>,
}
`;

exports[`Using lucide icon components should adjust the size, stroke color and stroke width 1`] = `
VueWrapper {
  "__app": Object {
    "_component": Object {
      "__emits": Object {},
      "__props": Array [
        Object {},
        Array [],
      ],
      "name": "VTU_ROOT",
      "render": [Function],
    },
    "_container": <div
      data-v-app=""
    >
      <svg
        fill="none"
        height="48"
        size="48"
        stroke="red"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        strokeWidth="4"
        viewBox="0 0 24 24"
        width="48"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="12"
          cy="12"
          r="10"
        />
        <path
          d="M8 14s1.5 2 4 2 4-2 4-2"
        />
        <line
          x1="9"
          x2="9.01"
          y1="9"
          y2="9"
        />
        <line
          x1="15"
          x2="15.01"
          y1="9"
          y2="9"
        />
      </svg>
    </div>,
    "_context": Object {
      "app": [Circular],
      "components": Object {
        "transition": Object {
          "name": "transition",
          "props": undefined,
          "render": [Function],
        },
        "transition-group": Object {
          "name": "transition-group",
          "props": undefined,
          "render": [Function],
        },
      },
      "config": Object {
        "errorHandler": undefined,
        "globalProperties": Object {},
        "isCustomElement": [Function],
        "isNativeTag": [Function],
        "optionMergeStrategies": Object {},
        "performance": false,
        "warnHandler": undefined,
      },
      "directives": Object {},
      "mixins": Array [
        Object {
          "__emits": null,
          "__props": Array [],
          "beforeCreate": [Function],
        },
      ],
      "provides": Object {},
      "reload": [Function],
    },
    "_props": null,
    "_uid": 1,
    "component": [Function],
    "config": Object {
      "errorHandler": undefined,
      "globalProperties": Object {},
      "isCustomElement": [Function],
      "isNativeTag": [Function],
      "optionMergeStrategies": Object {},
      "performance": false,
      "warnHandler": undefined,
    },
    "directive": [Function],
    "mixin": [Function],
    "mount": [Function],
    "provide": [Function],
    "unmount": [Function],
    "use": [Function],
    "version": "3.0.6",
  },
  "__setProps": [Function],
  "componentVM": Object {
    "hasOwnProperty": [Function],
  },
  "rootVM": Object {},
  "wrapperElement": <svg
    fill="none"
    height="48"
    size="48"
    stroke="red"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    strokeWidth="4"
    viewBox="0 0 24 24"
    width="48"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      cx="12"
      cy="12"
      r="10"
    />
    <path
      d="M8 14s1.5 2 4 2 4-2 4-2"
    />
    <line
      x1="9"
      x2="9.01"
      y1="9"
      y2="9"
    />
    <line
      x1="15"
      x2="15.01"
      y1="9"
      y2="9"
    />
  </svg>,
}
`;

exports[`Using lucide icon components should render an component 1`] = `
VueWrapper {
  "__app": Object {
    "_component": Object {
      "__emits": Object {},
      "__props": Array [
        Object {},
        Array [],
      ],
      "name": "VTU_ROOT",
      "render": [Function],
    },
    "_container": <div
      data-v-app=""
    >
      <svg
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="12"
          cy="12"
          r="10"
        />
        <path
          d="M8 14s1.5 2 4 2 4-2 4-2"
        />
        <line
          x1="9"
          x2="9.01"
          y1="9"
          y2="9"
        />
        <line
          x1="15"
          x2="15.01"
          y1="9"
          y2="9"
        />
      </svg>
    </div>,
    "_context": Object {
      "app": [Circular],
      "components": Object {
        "transition": Object {
          "name": "transition",
          "props": undefined,
          "render": [Function],
        },
        "transition-group": Object {
          "name": "transition-group",
          "props": undefined,
          "render": [Function],
        },
      },
      "config": Object {
        "errorHandler": undefined,
        "globalProperties": Object {},
        "isCustomElement": [Function],
        "isNativeTag": [Function],
        "optionMergeStrategies": Object {},
        "performance": false,
        "warnHandler": undefined,
      },
      "directives": Object {},
      "mixins": Array [
        Object {
          "__emits": null,
          "__props": Array [],
          "beforeCreate": [Function],
        },
      ],
      "provides": Object {},
      "reload": [Function],
    },
    "_props": null,
    "_uid": 0,
    "component": [Function],
    "config": Object {
      "errorHandler": undefined,
      "globalProperties": Object {},
      "isCustomElement": [Function],
      "isNativeTag": [Function],
      "optionMergeStrategies": Object {},
      "performance": false,
      "warnHandler": undefined,
    },
    "directive": [Function],
    "mixin": [Function],
    "mount": [Function],
    "provide": [Function],
    "unmount": [Function],
    "use": [Function],
    "version": "3.0.6",
  },
  "__setProps": [Function],
  "componentVM": Object {
    "hasOwnProperty": [Function],
  },
  "rootVM": Object {},
  "wrapperElement": <svg
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      cx="12"
      cy="12"
      r="10"
    />
    <path
      d="M8 14s1.5 2 4 2 4-2 4-2"
    />
    <line
      x1="9"
      x2="9.01"
      y1="9"
      y2="9"
    />
    <line
      x1="15"
      x2="15.01"
      y1="9"
      y2="9"
    />
  </svg>,
}
`;
